File: sspSparse.c
/**
Michael Motro github.com/motrom/fastmurty 4/2/19
*/
#ifdef SPARSE

#include "sspSparse.h"
#include <stddef.h> // size_t

WorkvarsforSSP allocateWorkvarsforSSP(int m, int n) {
	WorkvarsforSSP workvars;
    size_t totalsize = sizeof(Pathtype) * (m*n+m) +
                       sizeof(int) * n;
	workvars.Q = (Pathtype *) malloc(totalsize);
	workvars.pathback = (int *)(workvars.Q + m*n+m);
	workvars.m = m;
	workvars.n = n;
	return workvars;
};

void deallocateWorkvarsforSSP(WorkvarsforSSP workvars) {
	free(workvars.Q);
};


/* qPush and qPop are a standard binary heap push and pop
    unlike the interval heap of queue.c */
void qPush(Pathtype* Q, int pos, double newval, int newi, int newj){
    int parentpos;
    while (pos > 0){
        parentpos = (pos - 1) >> 1;
        if (newval > Q[parentpos].val){
            break;
        }
        Q[pos] = Q[parentpos];
        pos = parentpos;
    }
    Q[pos].val = newval;
    Q[pos].i = newi;
    Q[pos].j = newj;
};

Pathtype qPop(Pathtype* Q, int Qsize){
    Pathtype minele, newele;
    double newkey;
    int pos, childpos, rightpos;
    minele = Q[0];
    newele = Q[Qsize];
    newkey = newele.val;
    pos = 0;
    childpos = 1;
    while (childpos < Qsize){
        rightpos = childpos + 1;
        if (rightpos < Qsize){
            if (Q[childpos].val > Q[rightpos].val){
                childpos = rightpos;
            }
        }
        if (Q[childpos].val > newkey){
            break;
        }
        Q[pos] = Q[childpos];
        pos = childpos;
        childpos = (pos << 1) + 1;
    }
    Q[pos] = newele;
    return minele;
};

double SSP(cs_di c, Subproblem* prb, WorkvarsforSSP* workvars){
    const double inf = 1000000000;
	double C = 0;
	int cj, ri, i, i1, j, j1, minj, m, cstartidx, cendidx, Qsize;
    double minval, val, ui;
	Solution sol = prb->solution;
	Pathtype minpath;
	Pathtype* Q;
	int* pathback;
	#ifndef NDEBUG
	const int loopescape = 10000;
	int loopcounter = 0;
	#endif
	
	ui = 0;
	
	pathback = workvars->pathback;
	Q = workvars->Q;

    // reset y, v, pathback
    // cols2use indexing and the row-sparse matrix don't play well together
    // so inclusion information is handled by pathback
    for(j = 0; j < c.n; j++){
        pathback[j] = 0;
    }
    for(cj = 0; cj < prb->n; cj++){
        j = prb->cols2use[cj];
        sol.y[j] = -1;
        sol.v[j] = 0;
        pathback[j] = -2;
    }


    // basic column reduction - basically running some rows in a convenient order
    m = prb->m;
    for(ri=prb->m-1; ri>=0; ri--){
        i = prb->rows2use[ri];
        minval = 0;
        minj = -1;
        cstartidx = c.p[i];
        cendidx = c.p[i+1];
        for(cj=cstartidx; cj<cendidx; cj++){
            j = c.i[cj];
            if(pathback[j] == -2){
                val = c.x[cj];
                if(val < minval){
                    minval = val;
                    minj = j;
                }
            }
        }
        if((minj==-1) || (sol.y[minj]==-1)){
            // this row can be matched without conflicting previous matches
            sol.x[i] = minj;
			if (minj != -1) {
				sol.y[minj] = i;
				C += minval;
			}
            m--;
            prb->rows2use[ri] = prb->rows2use[m];
            prb->rows2use[m] = i;
        }
    }

    for(ri=0; ri<m; ri++){
		i1 = prb->rows2use[ri];
		Qsize = 0; // restart queue
		for(cj = 0; cj < prb->n; cj++){
            pathback[prb->cols2use[cj]] = -2;
        }
        // shortest paths algorithm
		cstartidx = c.p[i1];
		cendidx = c.p[i1+1];
        for(cj=cstartidx; cj<cendidx; cj++){
            j = c.i[cj];
            if (pathback[j] == -2){
                qPush(Q, Qsize, c.x[cj] - sol.v[j], i1, j);
                Qsize++;
            }
        }
        qPush(Q, Qsize, 0, i1, -1);
        Qsize++;
		for (;;) {
			for (;;) {
				assert(loopcounter++ < loopescape);
			    if (Qsize == 0){
			        return inf;
			    }
			    Qsize--;
			    minpath = qPop(Q, Qsize);
			    j = minpath.j;
			    if (j == -1) break; // hit unmatched row
	            if (pathback[j] == -2) break;
			}
			i = minpath.i;
			minval = minpath.val;
			if (j == -1) break;
			// this column should no longer be considered
			pathback[j] = i;
			sol.v[j] += minval;
			i = sol.y[j];
			if (i == -1) break;
			// update distances to other columns
			cstartidx = c.p[i];
			cendidx = c.p[i+1];
			// have to look up the right column for ui
			for (cj=cstartidx; cj<cendidx; cj++){
			    if (c.i[cj] == j){
			        ui = c.x[cj] - sol.v[j];
			    }
			}
			qPush(Q, Qsize, -ui, i, -1);
			Qsize++;
			for (cj=cstartidx; cj<cendidx; cj++){
				j = c.i[cj];
				if (pathback[j] == -2){
				    val = c.x[cj] - sol.v[j] - ui;
				    qPush(Q, Qsize, val, i, j);
				    Qsize++;
				}
			}
		}
		// travel back through shortest path
		if (j == -1) {
			j = sol.x[i];
			sol.x[i] = -1;
		}
		while (i != i1) {
			assert(loopcounter++ < loopescape);
			i = pathback[j];
			sol.y[j] = i;
			j1 = j;
			j = sol.x[i];
			sol.x[i] = j1;
		}
		// update reductions
		for (cj = 0; cj < prb->n; cj++) {
		    j = prb->cols2use[cj];
		    if (pathback[j] != -2){
			    sol.v[j] -= minval;
			}
		}
        
		// update total cost
		C += minval;
	}
	#ifndef NDEBUG
	double eps_debug = 0.0000001;
	for(j=0; j<workvars->n; j++){
	    pathback[j] = 0;
	}
	for(cj=0; cj<prb->n; cj++){
	    j = prb->cols2use[cj];
	    pathback[j] = 1;
	}
	for (ri=0; ri<prb->m; ri++){
	    i = prb->rows2use[ri];
	    j = sol.x[i];
	    cstartidx = c.p[i];
	    cendidx = c.p[i+1];
	    if (j==-1){
		    // check for positive slack on miss row
            for(cj = cstartidx; cj < cendidx; cj++){
                j1 = c.i[cj];
			    assert(sol.y[j1] != i);
			    if (pathback[j1]){
			        assert(c.x[cj] - sol.v[j1] > -eps_debug);
			    }
            }
	    } else {
	        // check for negative reduction and positive slack on match row
		    assert(sol.y[j] == i);
		    for(cj=cstartidx; cj<cendidx; cj++){
		        j1 = c.i[cj];
		        if (c.i[cj] == j){
		            ui = c.x[cj] - sol.v[j];
		        }
		    }
		    assert(ui < eps_debug);
		    for (cj=cstartidx; cj<cendidx; cj++) {
		        j1 = c.i[cj];
			    if ((j1 != j) & pathback[j1]){
			        assert(c.x[cj] - sol.v[j1] - ui > -eps_debug);
			    }
		    }
	    }
	}
	#endif
	return C;
};

double spStep(cs_di c, Subproblem* prb, WorkvarsforSSP* workvars, double cost_bound) {
	const double inf = 1000000000;
	int cj, ri, Qsize, i, i1, j, j1, cstartidx, cendidx;
	int minmissi = 0;//-1;
	int minmissj = 0;//i1;
	double minval, val, ui, missing_cost;
	bool missing, miss_unused, missing_from_row;
	Solution sol = prb->solution;
	Pathtype minpath;
	Pathtype* Q;
	int* pathback;
	#ifndef NDEBUG
    const int loopescape = 1000;
    int loopcounter = 0;
	#endif

    // cols2use indexing the row-sparse matrix don't play well together
    // so inclusion information is handled by pathback
    Q = workvars->Q;
    Qsize = 0;
    pathback = workvars->pathback;
    for(j = 0; j < c.n; j++){
        pathback[j] = 0;
    }
    for(cj = 0; cj < prb->n; cj++){
        pathback[prb->cols2use[cj]] = -2;
    }

	// which row and column are to be rematched
	i1 = prb->rows2use[prb->m-1];
	j1 = sol.x[i1];

	cstartidx = c.p[i1];
	cendidx = c.p[i1+1];
	// u not necessary to get solution, but gives accurate cost change
	ui = 0;
	for (cj=cstartidx; cj<cendidx; cj++){
	    if (c.i[cj] == j1){
	        ui = c.x[cj] - sol.v[j1];
	    }
	}
    for(cj=cstartidx; cj<cendidx; cj++){
        j = c.i[cj];
        if ((pathback[j] == -2) & (!prb->eliminateels[j])){
            val = c.x[cj] - sol.v[j] - ui;
            if (val < cost_bound){
                qPush(Q, Qsize, val, i1, j);
                Qsize++;
            }
        }
    }
    if (!prb->eliminatemiss){
        val = -ui;
        if (val < cost_bound){
            qPush(Q, Qsize, val, i1, -1);
            Qsize++;
        }
    }
	miss_unused = true;
	missing_from_row = false;
	missing_cost = 0; // this is a dual cost on auxiliary columns
	for (;;) {
		missing = false;
		for (;;) {
			assert(loopcounter++ < loopescape);
		    if (Qsize == 0){
		        return inf; // early stopping
		    }
		    Qsize--;
		    minpath = qPop(Q, Qsize);
		    if (minpath.j == -1){
		        if (miss_unused){
		            minmissi = minpath.i;
		            missing = true;
		            missing_from_row = true;
		            break; // hit unmatched row
		        }
            }
            if (pathback[minpath.j] == -2) break;
		}
		minval = minpath.val;
		if (!missing){
		    i = minpath.i;
		    j = minpath.j;
		    // this column should no longer be considered
		    pathback[j] = i;
		    sol.v[j] += minval;
		    if (j == j1) {
			    break;
		    }
		    i = sol.y[j];
			if (i == -1) {
				// entry to missing zone : col was missing but is now matched
				if (miss_unused) {
					minmissj = j;
					missing = true;
					missing_from_row = false;
				}
				else {
					// already covered the missing zone, this is a dead end
					continue;
				}
			}
		}
		if (missing) {
			if (j1 == -1) {
				j = -1;
				break;
			}
			miss_unused = false;
			missing_cost = minval;
			ui = -minval;
			// exit from missing zone : row that was missing is matched
			for (ri = 0; ri < prb->m; ri++) {
				i = prb->rows2use[ri];
				if (sol.x[i] == -1) {
					cstartidx = c.p[i];
					cendidx = c.p[i+1];
					for (cj = cstartidx; cj < cendidx; cj++) {
						j = c.i[cj];
						if (pathback[j]==-2){
						    val = c.x[cj] - sol.v[j] - ui;
						    if (val < cost_bound) {
							    qPush(Q, Qsize, val, i, j);
							    Qsize++;
						    }
						}
					}
				}
			}
			// exit from missing zone : col that was matched is missing
			for (cj = 0; cj < prb->n; cj++) {
				j = prb->cols2use[cj];
				if ((sol.y[j] != -1) & (pathback[j]==-2)) {
					val = -sol.v[j] - ui;
					if (val < cost_bound) {
						qPush(Q, Qsize, val, -1, j);
					    Qsize++;
					}
				}
			}
		}
		else {
		    // update distances to other columns
		    cstartidx = c.p[i];
			cendidx = c.p[i+1];
			// have to look up the right column for ui
			for (cj=cstartidx; cj<cendidx; cj++){
			    if (c.i[cj] == j){
			        ui = c.x[cj] - sol.v[j];
			    }
			}
			if (miss_unused){
			    val = -ui;
			    if (val < cost_bound){
			        qPush(Q, Qsize, -ui, i, -1);
			        Qsize++;
			    }
			}
			for (cj = cstartidx; cj < cendidx; cj++) {
				j = c.i[cj];
				if (pathback[j]==-2){
				    val = c.x[cj] - sol.v[j] - ui;
				    if (val < cost_bound) {
					    qPush(Q, Qsize, val, i, j);
					    Qsize++;
				    }
				}
			}
		}
	}
	// augment
	// travel back through shortest path to find matches
	i = i1 + 1; // any number that isn't i1
	while (i != i1) {
		assert(loopcounter++ < loopescape);
		if (j == -1) {
			// exit from missing zone : row was missing but is now matched
			i = -1;
		} else {
			i = pathback[j];
			sol.y[j] = i;
		}
		if (i == -1) {
			// exit from missing zone : column j was matched but is now missing
			if (missing_from_row) {
				// entry to missing zone : row was matched but is now missing
				i = minmissi;
				j = sol.x[i];
				sol.x[i] = -1;
			} else {
				// entry to missing zone : col was missing but is now matched
				j = minmissj;
			}
		} else {
			j1 = j;
			j = sol.x[i];
			sol.x[i] = j1;
		}
	}
	// updating of column reductions
	double updatevalchosen, updatevalnot;
	if (miss_unused) {
	    updatevalchosen = -minval;
	} else {
	    updatevalchosen = -missing_cost;
	}
	updatevalnot = minval + updatevalchosen;
	
	for (cj = 0; cj < prb->n; cj++) {
		j = prb->cols2use[cj];
        if (sol.y[j] == -1){
            sol.v[j] = 0;
        } else {
            if (pathback[j] == -2){
		        sol.v[j] += updatevalnot;
		    } else {
		        sol.v[j] += updatevalchosen;
		    }
        }
	}
	#ifndef NDEBUG
	    double eps_debug = 0.0000001;
	    bool eliminatingrow;
	    for(j=0; j<c.n; j++){
	        pathback[j] = 0;
	    }
	    for(cj=0; cj<prb->n; cj++){
	        j = prb->cols2use[cj];
	        pathback[j] = 1;
	    }
	    for (ri=0; ri<prb->m; ri++){
	        i = prb->rows2use[ri];
	        eliminatingrow = (i == i1);
	        j = sol.x[i];
	        cstartidx = c.p[i];
	        cendidx = c.p[i+1];
	        if (j==-1){
		        // check for positive slack on miss row
                for(cj = cstartidx; cj < cendidx; cj++){
                    j1 = c.i[cj];
			        assert(sol.y[j1] != i);
			        if (pathback[j1] & (!(eliminatingrow & prb->eliminateels[j1]))){
			            assert(c.x[cj] - sol.v[j1] > -eps_debug);
			        }
                }
	        } else {
	            // check for negative reduction and positive slack on match row
		        assert(sol.y[j] == i);
		        for(cj=cstartidx; cj<cendidx; cj++){
		            if (c.i[cj] == j){
		                ui = c.x[cj] - sol.v[j];
		            }
		        }
		        if(!(eliminatingrow & prb->eliminatemiss)){
		            assert(ui < eps_debug);
		        }
		        for (cj=cstartidx; cj<cendidx; cj++) {
		            j1 = c.i[cj];
			        if ((j1 != j) & pathback[j1] & (!(eliminatingrow & prb->eliminateels[j1]))){
			            assert(c.x[cj] - sol.v[j1] - ui > -eps_debug);
			        }
		        }
	        }
	    }
	#endif
	return minval;
};

#endif

File: previous python implementation/sspSparse.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
The Jonker-Volgenant algorithm for finding the maximum assignment.

Michael Motro, University of Texas at Austin
last modified 3/5/2019

the __main__ code at the bottom tests this implementation, comparing it to
Scipy's linear_sum_assignment function. You'll need to have scipy in your
distribution to run this file on its own, but not to import it in other files.
"""
import numpy as np
import numba as nb
from sparsity import nbsparsedtype

inf= 1e9 # inf is a suitably large number


nbheapouttype = nb.typeof((0., 0, 0)) # float, int, int

heapdtype = np.dtype([('key', np.float64), ('i',np.int64), ('j',np.int64)])
nbheapdtype = nb.from_dtype(heapdtype)

@nb.jit(nb.void(nbheapdtype[:], nb.i8, nb.f8, nb.i8, nb.i8), nopython=True)
def heappush(heap, pos, newkey, newi, newj):
    while pos > 0:
        parentpos = (pos - 1) >> 1
        if newkey > heap[parentpos]['key']: break
        heap[pos] = heap[parentpos]
        pos = parentpos
    inputel = heap[pos]
    inputel['key'] = newkey
    inputel['i'] = newi
    inputel['j'] = newj
    
@nb.jit(nbheapouttype(nbheapdtype[:], nb.i8), nopython=True)
def heappop(heap, heapsize):
    minele = heap[0]
    minkey = minele['key']
    mini = minele['i']
    minj = minele['j']
    heapsize -= 1
    newele = heap[heapsize]
    newkey = newele['key']
    # Bubble up the smaller child until hitting a leaf.
    pos = 0
    childpos = 1    # leftmost child position
    while childpos < heapsize:
        # Set childpos to index of smaller child.
        rightpos = childpos + 1
        if rightpos < heapsize:
            if heap[childpos]['key'] > heap[rightpos]['key']:
                childpos = rightpos
        if heap[childpos]['key'] > newkey:
            break
        # Move the smaller child up.
        heap[pos] = heap[childpos]
        pos = childpos
        childpos = (pos<<1) + 1
    heap[pos] = newele
    return minkey, mini, minj


@nb.jit(nb.f8(nbsparsedtype[:,:], nb.i8[:], nb.i8[:], nb.f8[:],
                nb.i8[:], nb.i8, nb.i8[:], nb.i8, nbheapdtype[:], nb.i8[:]),
        nopython=True)
def SSP(c, x, y, v, rows2use, nrows2use, cols2use, ncols2use, d, pred):
    """ solves full 2D assignment problem
    c: matrix, sparse
    x: column indices that match to row, or -1 if row is missing
    y: match indices for column
    v: column reductions
    rows2use, nrows2use: rows in rows2use[:nrows2use] are considered part of the problem
    cols2use, ncols2use: " "
    d: workspace var, heap structure to store potential matches (pathcost, i, j)
    pred: workspace var, stores whether each column was used and path backward
    returns cost of assignment
    """
    n = y.shape[0]
    C = 0.
    y[:] = -1
    v[:] = 0.
    pred[:] = 0
    
    # initial step
    pred[cols2use[:ncols2use]] = -2
    freerows = 0
    for ri in xrange(nrows2use):
        i = rows2use[ri]
        minj = -1
        minval = 0.
        for cij in c[i]:
            j = cij['idx']
            if pred[j] == -2:
                val = cij['x']
                if val < minval:
                    minval = val
                    minj = j
        if minj == -1:
            x[i] = -1
            rows2use[ri] = rows2use[freerows]
            rows2use[freerows] = i
            freerows += 1
        elif y[minj] == -1:
            x[i] = minj
            y[minj] = i
            C += minval
            rows2use[ri] = rows2use[freerows]
            rows2use[freerows] = i
            freerows += 1
    
    for i1 in rows2use[freerows:nrows2use]:
        dsize = 0 # restart heap
        # pred doubles as a check for previously visited columns, and a path backwards
        pred[cols2use[:ncols2use]] = -2
        rowi = c[i1]
        for cij in rowi:
            j = cij['idx']
            dj = cij['x'] - v[j]
            if pred[j] == -2:
                heappush(d, dsize, dj, i1, j)
                dsize += 1
        heappush(d, dsize, 0., i1, -1)
        dsize += 1
        while True:
            predj = 0
            while predj != -2: # already found and matched this col
                assert dsize > 0
                minval, i, j = heappop(d, dsize)
                dsize -= 1
                if j == -1:
                    break # hit unmatched row
                predj = pred[j]
            if j == -1: break
            pred[j] = i
            v[j] += minval # first half of score augmentation
            i = y[j]
            if i == -1:
                break # hit unmatched column
            # update distances to other columns
            rowi = c[i]
            # find this row's reduction
            # have to look up the right column
            for cij in rowi:
                if cij['idx']==j:
                    u1 = cij['x'] - v[j]
            heappush(d, dsize, -u1, i, -1)
            dsize += 1
            for cij in rowi:
                j = cij['idx']
                if pred[j] == -2:
                    dj = cij['x'] - v[j] - u1
                    heappush(d, dsize, dj, i, j)
                    dsize += 1
        
        # augment
        # travel back through shortest path to find matches
        if j==-1:
            j = x[i]
            x[i] = -1
        while i != i1:
            i = pred[j]
            y[j] = i
            k = j
            j = x[i]
            x[i] = k
        # updating of column prices, part 2
        for j in xrange(n):
            if pred[j] != -2:
                v[j] -= minval
        # updating total cost
        C += minval
    return C
                
                
            
@nb.jit(nb.f8(nbsparsedtype[:,:], nb.i8[:], nb.i8[:], nb.f8[:], nb.i8[:],
              nb.i8, nb.i8[:], nb.i8, nbheapdtype[:], nb.i8[:], nb.i8, nb.i8,
              nb.b1[:], nb.b1, nb.f8), nopython=True)
def spStep(c, x, y, v, rows2use, nrows2use, cols2use, ncols2use, d, pred,
           i1, j1, eliminate_els, eliminate_miss, cost_bound):
    """ solves Murty subproblem given solution to originating problem
    same inputs as SSP and also:
    i1, j1 = row and column that are now unassigned
    eliminate_els = boolean array, whether matching a column with i1 is prohibited
    eliminate_miss = whether i1 is prohibited to miss
    cost_bound = function will stop early and return inf if the solution is known
                    to be above this bound
    returns cost of shortest path, a.k.a. this solution's cost minus original solution's
    """
    pred[:] = 0
    pred[cols2use[:ncols2use]] = -2 # set these as available
    dsize = 0
    rowi = c[i1]
    ui = 0.
    for cij in rowi:
        if cij['idx'] == j1:
            ui = cij['x'] - v[j1]
    for cij in rowi:
        j = cij['idx']
        if pred[j] == -2 and not eliminate_els[j]:
            dj = cij['x'] - v[j] - ui
            if dj <= cost_bound:
                heappush(d, dsize, dj, i1, j)
                dsize += 1
    if not eliminate_miss:
        dj = -ui
        if dj <= cost_bound:
            heappush(d, dsize, dj, i1, -1)
            dsize += 1
    minmissi = 0
    minmissj = 0
    miss_unused = True
    missing_from_row = False
    missing_cost = 0. # this is a dual cost on auxiliary columns
    while True:
        missing = False
        predj = 0
        while predj != -2: # already found and matched this col
            if dsize == 0:
                return inf
            minval, i, j = heappop(d, dsize)
            dsize -= 1
            if j == -1:
                if miss_unused:
                    minmissi = i
                    missing = True
                    missing_from_row = True
                    break # hit unmatched row
            else:
                predj = pred[j]
        if not missing:
            pred[j] = i
            v[j] += minval # first half of score augmentation
            if j==j1: break
            i = y[j]
            if i==-1:
                # entry to missing zone: col was missing but is now matched
                if miss_unused:
                    minmissj = j
                    missing = True
                    missing_from_row = False
                else:
                    # already covered the missing zone, this is a dead end
                    continue
        if missing:
            if j1 == -1:
                j=-1
                break
            miss_unused = False
            missing_cost = minval
            u1 = -minval
            # exit from missing zone: row that was missing is matched
            for i in rows2use[:nrows2use]:
                if x[i]==-1:
                    rowi = c[i]
                    for cij in rowi:
                        j = cij['idx']
                        if pred[j] == -2:
                            dj = cij['x']-v[j]-u1
                            if dj <= cost_bound:
                                heappush(d, dsize, dj, i, j)
                                dsize += 1
            # exit from missing zone: col that was matched is missing
            for j in cols2use[:ncols2use]:
                if y[j] >= 0 and pred[j] == -2:
                    dj = -v[j]-u1
                    if dj <= cost_bound:
                        heappush(d, dsize, dj, -1, j)
                        dsize += 1
        else:
            rowi = c[i]
            for cij in rowi:
                # first need to find this row's price
                # meaning we have to look up the right column
                if cij['idx']==j:
                    ui = cij['x'] - v[j]
            if miss_unused:
                dj = -ui
                if dj <= cost_bound:
                    heappush(d, dsize, dj, i, -1)
                    dsize += 1
            for cij in rowi:
                j = cij['idx']
                if pred[j] == -2:
                    dj = cij['x'] - v[j] - ui
                    if dj <= cost_bound:
                        heappush(d, dsize, dj, i, j)
                        dsize += 1
    
    # augment
    # travel back through shortest path to find matches
    i = i1+1 # any number that isn't i1
    while i != i1:
        if j == -1:
            # exit from missing zone: row was missing but is now matched
            i = -1
        else:
            i = pred[j]
            y[j] = i
        if i == -1:
            # exit from missing zone: column j was matched but is now missing
            if missing_from_row:
                # entry to missing zone: row was matched but is now missing
                i = minmissi
                j = x[i]
                x[i] = -1
            else:
                # entry to missing zone: col was missing but is now matched
                j = minmissj
        else:
            k = j
            j = x[i]
            x[i] = k
    # updating of column prices
    for j in cols2use[:ncols2use]:
        if pred[j]!=-2:
            v[j] -= minval
    if not miss_unused:
        v[cols2use[:ncols2use]] += minval - missing_cost
    v[y==-1] = 0.
    return minval




if __name__ == '__main__':
    """
    create a random matrix
    try assignment, check for equality
    """
    from scipy.optimize import linear_sum_assignment
    from sparsity import sparsedtype
    
    m = 10
    n = 20
    s = 10
#    P = np.random.exponential(size=(n,m))
#    mX = np.random.exponential(size=(n,))
#    mY = np.random.exponential(size=(m,))
    P = np.random.rand(m,n)*5
    mX = np.random.rand(m)
    mY = np.random.rand(n)
    
    # make sparse version
    # take s lowest columns from each row
    c2 = P - mX[:,None] - mY[None,:]
    csp = np.zeros((m,s), dtype=sparsedtype)
    for i in xrange(m):
        colsinorder = np.argsort(c2[i])
        csp[i]['idx'] = colsinorder[:s]
        csp[i]['x'] = c2[i, colsinorder[:s]]
        # make it sparse on the real matrices too
        P[i,colsinorder[s:]] = 100
        c2[i,colsinorder[s:]] = .01
    
    # make full square version, use standard code
    c1 = np.zeros((m+n,m+n))
    c1[:m,:n] = P
    c1[:m,n:] = 1e4
    c1[range(m),range(n,m+n)] = mX
    c1[m:,:n] = 1e4
    c1[range(m,m+n), range(n)] = mY
    sol = linear_sum_assignment(c1)
    x1 = np.array(sol[1][:m])
    x1[x1>=n] = -1
    y1 = np.arange(n)
    for k,j in enumerate(sol[1]):
        j = sol[1][k]
        if j < n:
            if k < m:
                y1[j] = k
            else:
                y1[j] = -1
    print(x1)
    print(y1)

    y = np.zeros(n, dtype=int) - 1
    x = np.zeros(m, dtype=int) - 1
    v = np.zeros(n)
    
    rows2use = np.arange(m)
    cols2use = np.arange(n)
    d = np.zeros(m*s+m, dtype=heapdtype)
    pred = np.zeros(n, dtype=int)
    C = SSP(csp, x, y, v, rows2use, cols2use, d, pred)
    print(x)
    print(y)
    
    v += mY
    u = mX.copy()
    xmatch = x>=0
    xmis = xmatch==False
    ymis = y==-1
    u[xmatch] = P[xmatch,x[xmatch]] - v[x[xmatch]]
    u2 = np.append(u, np.zeros(n))
    v2 = np.append(v, np.zeros(m))
    x2 = np.append(x, y+n)
    x2[np.where(x==-1)[0]] = np.where(x==-1)[0]+n
    x2[np.where(y==-1)[0]+m] = np.where(y==-1)[0]
    slack = c1 - u2[:,None] - v2
    assert np.min(slack) > -1e-8
    assert all(slack[range(m+n), x2] < 1e-8)
    assert np.min(v[ymis]) >= -1e-8 if any(ymis) else True
    C2 = sum(c2[i,j] for i,j in enumerate(x) if j>=0)
    assert np.isclose(C, C2)